// @soundworks/core/server/AbstractPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _debug = _interopRequireDefault(require("debug"));

var _lodash = _interopRequireDefault(require("lodash.merge"));

var _Signal = _interopRequireDefault(require("../common/Signal.js"));

var _SignalAll = _interopRequireDefault(require("../common/SignalAll.js"));

var _logger = _interopRequireDefault(require("./utils/logger.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const log = (0, _debug.default)('soundworks:lifecycle');
/**
 * Base class to extend for creating new soundworks plugins.
 *
 * @memberof server
 */

class AbstractPlugin {
  constructor(server, name) {
    // should receive soundworks instance as argument

    /**
     * Instance of soundworks server.
     * @type {server.Server}
     * @see {@link server.Server}
     */
    this.server = server;
    /**
     * Name of the plugin.
     * @type {String}
     */

    this.name = name;
    /**
     * Options of the plugin.
     * @type {Object}
     */

    this.options = {};
    /**
     * Signals defining the process state.
     * @type {Object}
     */

    this.signals = {
      start: new _SignalAll.default(),
      started: new _Signal.default(),
      ready: new _Signal.default(),
      errored: new _Signal.default()
    };
    /** @private */

    this.clientTypes = new Set(); // register in the server

    this.server.activities.add(this); // start when all required signals are fired

    this.signals.start.addObserver(value => {
      _logger.default.pluginStart(this.name);

      this.start();
    }); // log errors if any

    this.signals.errored.addObserver(value => _logger.default.pluginErrored(this.name));
    this.ready = this.ready.bind(this);
  }
  /**
   * Helper function to merge default options and user-defined options.
   *
   * @example
   * this.options = this.configure(defaults, options);
   */


  configure(defaults, options) {
    return (0, _lodash.default)({}, defaults, options);
  }
  /**
   * Interface method to override when implementing child class.
   *
   * _The child class MUST call {@link server.AbstractPlugin#started} and
   * {@link server.AbstractPlugin#ready}_
   *
   * @example
   * class MyDelayPlugin extends AbstractPlugin {
   *   // start() is executed when the `start` signal pass to `true`
   *   async start() {
   *     this.state = await this.server.stateManager.create(`s:${this.name}`);
   *     this.started();
   *     // do [async] stuff
   *     setTimeout(() => this.ready(), 3000);
   *   }
   * }
   */


  start() {
    throw new Error(`plugin "${this.name}": "start()" not implemented or "super.start()" called`);
  }
  /**
   * Method to call when the plugin is effectively started, as it may have to
   * do some asynchonous job at start (e.g. creating a shared state).
   *
   * _Must be called between before {@link server.AbstractPlugin#ready}._
   *
   * @example
   * class MyDelayPlugin extends AbstractPlugin {
   *   // start() is executed when the `start` signal pass to `true`
   *   async start() {
   *     this.state = await this.server.stateManager.create(`s:${this.name}`);
   *     this.started();
   *     // do [async] stuff
   *     setTimeout(() => this.ready(), 3000);
   *   }
   * }
   */


  started() {
    // @note - these check are mostly there for development help. Could be
    // replaced with decorators.
    if (this.signals.start.value === false) {
      throw new Error(`plugin "${this.name}" cannot "started" before "start"`);
    }

    if (this.signals.started.value === true) {
      throw new Error(`plugin "${this.name}" cannot be "started" twice`);
    }

    _logger.default.pluginStarted(this.name);

    this.signals.started.value = true;
  }
  /**
   * Method to call in the plugin lifecycle when it should be considered as
   * ready and thus allow the initialization process to continue or the
   * application to start safely (cf. {@link server.AbstractExperience#start}).
   *
   * @example
   * class MyDelayPlugin extends AbstractPlugin {
   *   // start() is executed when the `start` signal pass to `true`
   *   async start() {
   *     this.state = await this.server.stateManager.create(`s:${this.name}`);
   *     this.started();
   *     // do [async] stuff
   *     setTimeout(() => this.ready(), 3000);
   *   }
   * }
   */


  ready() {
    // @note - these check are mostly there for development help. Could be
    // replaced with decorators.
    if (this.signals.start.value === false) {
      throw new Error(`plugin "${this.name}" cannot "ready" before "start"`);
    }

    if (this.signals.started.value === false) {
      throw new Error(`plugin "${this.name}" cannot "ready" before "started"`);
    }

    if (this.signals.ready.value === true) {
      throw new Error(`plugin "${this.name}" cannot be "ready" twice`);
    }

    _logger.default.pluginReady(this.name);

    this.signals.ready.value = true;
  }

  error(msg) {
    console.log(`> plugin "${this.name}" error: ${msg}`);
    this.signals.errored.msg = msg;
    this.signals.errored.value = true;
  }
  /**
   * Interface method to implement if specific logic should be done when a
   * {@link server.Client} connects to the plugin.
   *
   * @param {server.Client} client
   */


  connect(client) {
    log(`> [client ${client.id}] connect to plugin "${this.name}"`);
  }
  /**
   * Interface method to implement if specific logic should be done when a
   * {@link server.Client} disconnects from the plugin.
   *
   * @param {server.Client} client
   */


  disconnect(client) {
    log(`> [client ${client.id}] disconnect from plugin "${this.name}"`);
  }

}

var _default = AbstractPlugin;
exports.default = _default;
